### ISSUE TITLE
Moving GPUParticles2D with collision enabled causes particles to collide with nothing

### ISSUE THREAD

**Issue by NewDefectus** [godotengine/godot#77553]
### Godot version

4.1.dev3

### System information

Windows 10, Forward+, Nvidia GeForce GTX 1050

### Issue description

When a GPUParticles2D instance that has collision detection enabled and local coords disabled is moved (or its parent is moved, etc.), it appears that some of the particles it has generated will sometimes bounce as if the light occluders moved for them.

### Steps to reproduce

The minimal reproduction project has a GPUParticles2D node that emits particles to the right, and a rectangular LightOccluder2D next to it. Pressing "1" moves the particles node to the left of the occluder, and pressing "2" moves it to its right.

It's expected that all particles on the right of the occluder will always move right (as there's nothing occluding them). However, rapidly alternating between 1 and 2 in succession will cause some particles on the occluder's right to start moving left.

### Minimal reproduction project

[LocalParticlesTest.zip](https://github.com/godotengine/godot/files/11582633/LocalParticlesTest.zip)

**Comment by Calinou** [comment]
I can confirm this on 4.1.dev 06ccbfe8e (Linux, GeForce RTX 4090 with NVIDIA 530.41.03). Increasing **Fixed FPS** to 1000 or decreasing it to 0 does not resolve this issue.

**Comment by NewDefectus** [comment]
Confirmed for 4.2.dev1 and improved the minimal reproduction project. I'd be happy if this bug were prioritized higher as this is affecting some of the effects in my game.

**Comment by Calinou** [comment]
> I'd be happy if this bug were prioritized higher as this is affecting some of the effects in my game.

We don't know what is causing this issue, and contributors work on a best-effort basis.

**Comment by WagnerGFX** [comment]
I tried running the minimal reproduction project on 4.2 beta1 and didn't notice any issue. I even changed a few settings, but the collisions didn't show any strange behavior.

It might have been fixed already.

**Comment by NewDefectus** [comment]
> I tried running the minimal reproduction project on 4.2 beta1 and didn't notice any issue. I even changed a few settings, but the collisions didn't show any strange behavior.
> 
> It might have been fixed already.

I checked again now and it seems that the bug is still occurring on my end. For reference here's the kind of mouse movement I meant:


https://github.com/godotengine/godot/assets/36775845/91a76079-b6e3-448a-b735-58c659ecd25e

**Comment by WagnerGFX** [comment]
> For reference here's the kind of mouse movement I meant:
> particles_bug.mp4

Oh yeah. Now I can reproduce the bug 👀 (I was moving too slow to notice)

**Comment by WagnerGFX** [comment]
> this is affecting some of the effects in my game.

I've discovered a workaround for the issue, at least for 4.2: Instead of moving the node, you can move the particle's emission_shape_offset and Visibility Rect.
With this the issue disappears, the downside is that process material needs to be unique for every particle node you use with this method.

> It's worth checking whether this issue occurs in 3D too.

I've also made a scene to test the issue in 3D. From what I'm seeing, it doesn't seem to happen in 3D.
Some particles may get stuck on the walls, but that's a separate issue.

Well... I was mistaken before, so check the project below to be sure. 😅

[LocalParticlesPlus.zip](https://github.com/godotengine/godot/files/13209625/LocalParticlesPlus.zip)

**Comment by RayoROAR** [comment]
> I've discovered a workaround for the issue, at least for 4.2: Instead of moving the node, you can move the particle's emission_shape_offset and Visibility Rect. With this the issue disappears, the downside is that process material needs to be unique for every particle node you use with this method.

Oh. My. God. You are a life saver! I stumbled upon the same problem and your solution works like a charm, many thanks!! :DD

**Comment by Blaine-McK** [comment]
> > this is affecting some of the effects in my game.
> 
> I've discovered a workaround for the issue, at least for 4.2: Instead of moving the node, you can move the particle's emission_shape_offset and Visibility Rect. With this the issue disappears, the downside is that process material needs to be unique for every particle node you use with this method.
> 
> > It's worth checking whether this issue occurs in 3D too.
> 
> I've also made a scene to test the issue in 3D. From what I'm seeing, it doesn't seem to happen in 3D. Some particles may get stuck on the walls, but that's a separate issue.
> 
> Well... I was mistaken before, so check the project below to be sure. 😅
> 
> [LocalParticlesPlus.zip](https://github.com/godotengine/godot/files/13209625/LocalParticlesPlus.zip)

I tried this solution with no luck unfortunately. I did download op's LocalParticles example and did implement your solution in it and it worked. I think my issue is that I am also rotating. The mouse position never rotates

**Comment by WagnerGFX** [comment]
I've checked this issue using v4.4-dev6, it still remains the same for 2D, while the 3D particles seem to be working perfectly fine.

> I think my issue is that I am also rotating.

From what I tested, it seems that any changes to the Particle2D's Transform will result in bugs. This includes moving, rotating and skewing.

> The mouse position never rotates

In that case, you need to compensate the offset position with the particle's rotation. This may require some vector calculations.

**Comment by qwoj274** [comment]
i solved it by duplicating child particle node, emitting it then removing (code attached to child particle node):

```
func _on_particle_emitting() -> void:
	var particle_copy: GPUParticles2D = duplicate()
	particle_copy.global_position = global_position
	particle_copy.top_level = true
	get_parent().add_child(particle_copy)
	particle_copy.emitting = true
	await particle_copy.finished
	particle_copy.queue_free()
```

_on_particle_emitting - some func connected to the signal that emits the particle.
duplicating particle node allows to emit particle even the previous emitting is not finished

**Comment by bennbollay** [comment]
Reproduced on v4.4.1.stable.official [49a5bc7b6] on MacBook Pro 15.6.1 (24G90).

**Comment by Joonnas** [comment]
This issue still persists on v4.6 rc2 [78c6632]
I have spent some time analyzing the code a bit. Here is what I found:

**TL;DR:** The SDF information is processed after the particles have been updated. Essentially what this means is that the collision information "lags" behind one frame. There are a few ideas to solve this (see below), but for now, I would suggest using the workarounds already mentioned in this discussion.


### The issue

In `RenderingServerDefault::_draw(...)` (rendering_server_default.cpp) we call `RSG::particles_storage->update_particles()` and afterwards `RSG::viewport->draw_viewports(...)`.
- In `update_particles` we call `_particles_process` in which we retrieve the SDF texture and information (e.g. sdf_collision_transform). This information has been stored in the particle system using the function `ParticlesStorage::particles_set_canvas_sdf_collision`
- In `draw_viewports` we iterate over each existing viewport and call `_draw_viewport`. Here we process the SDF (`if (p_viewport->sdf_active) {...}`). That means here we render the SDF texture and compute the SDF information.
	- Further down in `_draw_viewport` we go over each canvas and render it (`RSG::canvas->render_canvas(...)`). When tracing this down a bit, we arrive at the function `RendererCanvasCull::_render_canvas_item_tree` which calls `RendererCanvasCull::_cull_canvas_item`. Here we compute the final transforms (screen space positions) of our canvas items (and therefore also of our particle system).
	- Tracing the rendering function further down we eventually arrive at the function `RendererCanvasRenderRD::_record_item_commands(...)` which goes over each item command in a batch. In the `Item::Command::TYPE_PARTICLES` case we call `ParticlesStorage::particles_set_canvas_sdf_collision`. Here we update the SDF collision information on the particles.

This means that the particles work with the SDF / collision information from the frame before. The problem here is however not, that we have the SDF information of the viewport a frame too late (actually, this results in a collision where no collision should be / no collision where one should be when we move the collider, but it is quite difficult to reproduce this during actual gameplay and the visual bug of this is extremely minor). The major problem here is that the "sdf_collision_transform", which essentially is the screen space transform of our particle emitter, is computed and stored in the particles storage only during the rendering process.

With this insight it might also be that this issue is related to #86173.


### Potential solutions

Solving this problem is a bit more involved. We could:
1. Force an early SDF pass before particle update, and therefore decouple the SDF generation from the rendering stage. This would however only solve the creation of the SDF texture, not of the computation of the "sdf_collision_transform"
2. Fully decouple this problem from rendering which probably would involve to compute all the screen-space coordinates (including "sdf_collision_transform") not during the render call, but before the particles update. This would mean a major refactor and introduces a potentially worse software architecture.
3. Implement the particles update after the actual rendering. But then the rendering would lag behind.

These ideas are neither ideal nor am I 100% sure yet how or even if they could be implemented.

What we could also think about, is to implement a different particle collision system that resembles the way the 3D GPU particles do collisions, using special Attractor and Collision nodes, this is however probably a discussion for #60909.
